sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NonAsyncTaskShouldNotReturnNull.cs

```diff
@@ -0,0 +1,114 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class NonAsyncTaskShouldNotReturnNull : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4586";
+        private const string MessageFormat = "Do not return null from this method, instead return 'Task.FromResult<T>(null)', " +
+            "'Task.CompletedTask' or 'Task.Delay(0)'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static ISet<KnownType> TaskTypes =
+            new HashSet<KnownType>
+            {
+                KnownType.System_Threading_Tasks_Task,
+                KnownType.System_Threading_Tasks_Task_T
+            };
+
+        private static ISet<SyntaxKind> TrackedNullLiteralLocations =
+            new HashSet<SyntaxKind>
+            {
+                SyntaxKind.ArrowExpressionClause,
+                SyntaxKind.ReturnStatement,
+            };
+
+        private static ISet<SyntaxKind> PossibleEnclosingSyntax =
+            new HashSet<SyntaxKind>
+            {
+                SyntaxKind.VariableDeclaration,
+                SyntaxKind.PropertyDeclaration,
+                SyntaxKind.MethodDeclaration
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var nullLiteral = (LiteralExpressionSyntax)c.Node;
+
+                    if (!nullLiteral.Parent.IsAnyKind(TrackedNullLiteralLocations))
+                    {
+                        return;
+                    }
+
+                    var enclosingMember = nullLiteral.Ancestors().FirstOrDefault(n => n.IsAnyKind(PossibleEnclosingSyntax));
+                    if (enclosingMember == null ||
+                        enclosingMember.IsKind(SyntaxKind.VariableDeclaration))
+                    {
+                        return;
```

Comment:
I am sorry for not noticing this before, but I suspect we will not catch `return (null);`.

I not sure if it is worth adding support for:
```
return condition ? Task.FromResult(5) : null;
```